doc/developer/design/20220511_replicated_compute_introspection.md

```diff
@@ -0,0 +1,360 @@
+for clusters that only have a single replica. Only when more than one replica is
+present the query results become inconsistent, as queries might be answered by
+either of the replicas.
+
+While inconsistent query results are obviously undesirable, this approach
+involves the least implementation effort (none) and might be something we can
+accept for the time being, until the STORAGE infrastructure has been fleshed out
+more.
+
+### Merge Responses from All Replicas
+
+The compute controller can, instead of immediately returning the peek response
+of the fastest replica, wait for all responses and then return their union. This
+way, the collected introspection data is complete.
+
+The compute controller needs to decide which queries require a "peek-all"
+instead of the usual "peek-one", based on whether they read from introspection
+tables or not. However, "peek-all" queries will often return wrong results. For
+example:
+
+```sql
+SELECT count(*) FROM mz_materializations
+```
+
+... should return a single row, but would instead return one row for each replica
+response, if responses are merged via UNION.
+
+If replicas are slow to respond (e.g. because they are overloaded), the whole
+query becomes slow. If a replica is not responding at all, the compute
+controller waits forever and the query never returns. We could work around this
+by enforcing a timeout, but then we risk getting back a response that lacks data
+about the slowest (and therefore probably most interesting) replicas.
+
+### Replica-targeted Queries
+
+The compute controller can be extended to support sending compute commands to
+specific replicas only. This way, the response is guaranteed to originate from
+the target replica and any returned introspection data is consistent for that
+replica.
+
+With this approach, replicas are only able to answer introspection queries about
+themselves. If ADAPTER wants to provide a unified view over all replicas (and
+clusters) to the user, it has to explicitly query every replica and merge the
+results. This, again, causes issues if replicas are slow to respond or don’t
+respond at all.
+
+It is also not clear how introspection query capabilities should be made
+accessible to the user. We might need a new SQL construct like
+`SELECT … AT <replica>`.
+
+## Open questions
+
+* **How is the coordinator made aware of the storage location (blob URI, consensus
+  URI)?**
+
+  Should we pass this information via `--replicated-compute-introspection` for
+  now? At some point this needs to be solved properly, but maybe it's fine to
+  leave it out of scope here?
+
+* **What happens if all handles to a persist shard are dropped temporarily?**
+
+  This can happen if a replica is restarted and there are no concurrent readers.
+  Is the shard dropped? Is the replica able to open it again after the restart?
+  Do we need to keep a handle open in ADAPTER?
+  
+* **Does persist compaction work if there is only a single writer to a shard?**
```

Comment:
A persist shard's `since` is only ever advanced in response to read handles downgrading it. So yeah, something will have to be in charge of downgrading these sinces